modules/uid2IdSystem_shared.js

```diff
@@ -36,49 +146,154 @@ export class Uid2ApiClient {
+          }
+        } catch (_err) {
+          rejectPromise(responseText);
+        }
+      },
+      error: (error, xhr) => {
+        try {
+          this._logInfo('Error status, assuming unencrypted JSON');
+          const response = JSON.parse(xhr.responseText);
           const result = this.ResponseToRefreshResult(response);
           if (typeof result === 'string') { rejectPromise(result); } else { resolvePromise(result); }
-        } else {
-          this._logInfo('Decrypting refresh API response');
-          const encodeResp = this.createArrayBuffer(atob(req.responseText));
-          window.crypto.subtle.importKey('raw', this.createArrayBuffer(atob(refreshDetails.refresh_response_key)), { name: 'AES-GCM' }, false, ['decrypt']).then((key) => {
-            this._logInfo('Imported decryption key')
-            // returns the symmetric key
-            window.crypto.subtle.decrypt({
-              name: 'AES-GCM',
-              iv: encodeResp.slice(0, 12),
-              tagLength: 128, // The tagLength you used to encrypt (if any)
-            }, key, encodeResp.slice(12)).then((decrypted) => {
-              const decryptedResponse = String.fromCharCode(...new Uint8Array(decrypted));
-              this._logInfo('Decrypted to:', decryptedResponse);
-              const response = JSON.parse(decryptedResponse);
-              const result = this.ResponseToRefreshResult(response);
-              if (typeof result === 'string') { rejectPromise(result); } else { resolvePromise(result); }
-            }, (reason) => this._logWarn(`Call to UID2 API failed`, reason));
-          }, (reason) => this._logWarn(`Call to UID2 API failed`, reason));
+        } catch (_e) {
+          rejectPromise(error)
         }
-      } catch (err) {
-        rejectPromise(err);
       }
+    }, refreshDetails.refresh_token, { method: 'POST',
+      customHeaders: {
+        'X-UID2-Client-Version': this._clientVersion
+      } });
+    return promise;
+  }
+
+  async callCstgApi(request) {
+    this._logInfo('Building CSTG request for', request);
+    const box = await UID2CstgBox.build(stripPublicKeyPrefix(this._serverPublicKey))
+    const encoder = new TextEncoder();
+    const now = Date.now();
+    const { iv, ciphertext } = await box.encrypt(
+      encoder.encode(JSON.stringify(request)),
+      encoder.encode(JSON.stringify([now]))
+    );
+
+    const exportedPublicKey = await UID2CstgCrypto.exportPublicKey(box.clientPublicKey);
+    const requestBody = {
+      payload: UID2CstgCrypto.bytesToBase64(new Uint8Array(ciphertext)),
+      iv: UID2CstgCrypto.bytesToBase64(new Uint8Array(iv)),
+      public_key: UID2CstgCrypto.bytesToBase64(new Uint8Array(exportedPublicKey)),
+      timestamp: now,
+      subscription_id: this._subscriptionId,
     };
-    this._logInfo('Sending refresh request', refreshDetails);
-    req.send(refreshDetails.refresh_token);
+    return this._callCstgApi(requestBody, box)
```

Comment:
I think this should be awaited

Reply:
Coz we had `return` here which will return a promise, and I put await at where we call this function
`  const response = await client.callCstgApi(request)`

Reply:
TIL both work. I thought without the await it would return a promise of a promise.

Interestingly, ESLint says ["It can now be slower to remove `await` rather than keeping it"](https://eslint.org/docs/latest/rules/no-return-await).

Reply:
Thanks for pointing out, I don't know this before. I just found Prebid.js is still using eslint 7.29, which this has this `Redundant use of `await` on a return value` rule, let's keep it as this rather than disable the rule?

Reply:
Sounds good :)